test/include/test.hpp

```diff
@@ -402,6 +403,30 @@ auto within_abs(T px, U py, double ptol = 1e-6f)
         px, py, ptol);
 }
 
+template <class Iterator1, class Iterator2>
```

Comment:
(I have some overall concerns..)

1. Is this really globbing (which works on files) or is it regexp (which works on strings)? The api name vs what is inside makes me wonder.
2. If it is for regexp, we could use a standard libary maybe..? https://en.cppreference.com/w/cpp/regex. The code below is fairly complex, and I am wondering do we need this afresh ..? 
3. If this is a general purpose API, this shouldn't be in test.hpp
4. Handling of '*' can be fairly complex. I don't know how it is handling complex usage (because '*' would need some amount of backtracking etc..) 
5. One would need a unit test for this kind of api to test its functionality through various complexities of handling '*'. 
Thanks.

Reply:
> Is this really globbing (which works on files) or is it regexp (which works on strings)? The api name vs what is inside makes me wonder.

It is globbing on string. It is not a regex. Its similar to the [`fnmatch`](https://www.man7.org/linux/man-pages/man3/fnmatch.3.html) function.

> If it is for regexp, we could use a standard libary maybe..? https://en.cppreference.com/w/cpp/regex. The code below is fairly complex, and I am wondering do we need this afresh ..?

Regex in the C++ standard library is really slow(slower than calling to perl or python), which is why I didnt use it. Also globbing syntax is much simpler and its backwards compatible with the exact matching we were using. A regex most likely would need to be a seperate flag. 

> If this is a general purpose API, this shouldn't be in test.hpp

The test.hpp header is meant to be a standalone header file, which is why it wasn't added to our migraphx/algorithm.hpp header. 

> Handling of * can be fairly complex. I don't know how it is handling complex usage (because * would need some amount of backtracking etc..)

It uses recursion to handle the backtracking. 

> One would need a unit test for this kind of api to test its functionality through various complexities of handling *.

We dont have any unit tests for the test.hpp header, but some unit tests would be a good idea.

Reply:
The need for regexp (glob is the incorrect name here in this code, and should be fixed anyways..) looks like a very basic one that the c++ regexp library shouldn't be discounted without some comparative test results -- I am uncomfortable with the idea of home-brewed globbing/regexp.

Reply:
This is using a globbing expression or wildcard expression not a regular expression(which is why it is named as such). The globbing expression can be converted to a regex but it is not trivial conversion(see a C++ 17 translation [here](https://github.com/p-ranav/glob/blob/master/source/glob.cpp#L21)). Since we are only supporting a smaller subset(ie just `*` and `?`), its a fairly simple algorithm to implement glob matching directly, and I think much simpler than translating to a regex.

Reply:
Tests added for this function.

Reply:
PTAL at the sample code that I provided, just for comparison, without writing all the globbing stuff. We are doing string related operations here. Thanks.

Reply:
[external commentary] FWIW - I'm okay with this implementation if it meets your needs and prefer it to the regex suggestion because of the complexity of transforming a glob pattern to a regex pattern and making sure the escaping is happening correctly. The function needs a specification. The operation should also return the position of the last matched item (which makes the bool redundant).

Reply:
@sean-parent Thanks for the commentary.

> The function needs a specification

Can you explain more how the specification should be defined?

> The operation should also return the position of the last matched item (which makes the bool redundant).

I actually considered returning an iterator to the first mismatch when I was implementing this originally.(I dont think returning the last matched item would work for cases where the first element doesnt match). However, this is not as simple as returning `start`. 

After the call to `std::mismatch` it is the first mismatch unless its a glob. After the star loop, `start` will point to the iterator that did match, which seem ambiguous to return. I could try to keep track of the mismatches and return it but it would make the implementation a little more complicated.

I preferred to keep it simpler and just return a bool. We aren't going to use the iterator for this context, and if we need it in the future we could look at updating it.

Reply:
> [external commentary] FWIW - I'm okay with this implementation if it meets your needs and prefer it to the regex suggestion because of the complexity of transforming a glob pattern to a regex pattern and making sure the escaping is happening correctly. The function needs a specification. The operation should also return the position of the last matched item (which makes the bool redundant).

Thanks, @sean-parent. 

The representation of the glob pattern here is _trivial_. These are test names, not even file names, as far as I have seen. Tests so far I have seen are named as `[a-z|A-Z|0-9|_]+`. There was a hint somewhere in the tests that maybe '[' and ']' might also be present in these strings, but haven't found it. For all this basic stuff , a simple 6 liner escaping of all the regex meta characters is a good start, as shown in this code:  https://github.com/ROCm/AMDMIGraphX/pull/2545#discussion_r1442390538 . And of course, plenty to room to enhance this, as required.. BTW, as a backdrop, from performance point of view, this is simply about filtering (via regex or something else..) a list of maybe up to 250 test-names, and then running those tests -- running those tests takes waaay longer than just filtering these names. So tests might run for minutes, and the regex solution might take say 1 ms longer :-) 

Some of my thoughts I would like to share with you, as a reviewer:

Had some crazy complex pattern been presented in the problem statement, or in the course of review, I would be the first to say: yes, we should escape from the escape sequences. A sample test file presented is also trivial. 

I too first thought that maybe (as you point out) there is some need to look at the matching sub-patterns etc. Hence, I gave some sample regex code to try. But so far, no such requirement for sub-patterns has emerged. This is a rather simple use-case for std::regex.

The recommendation for using std::regex was given after I saw the attempted custom 'globbing' code here is simply a confusing use of std::mismatch -- with two meta characters `*` and `?` kicking control of there -- and then onto some recursive loopy stuff to take care of basic handling of string. IMHO, this kind of confusing stuff isn't required, when simple alternative exist in the standard library. I, as a reviewer, couldn't be sure looking at this custom code, exactly what is the requirement it is trying to fulfil.

Had a complex requirement been presented, definitely a custom solution can be justified. There is no such requirement, what-so-ever. The standard library offers such a rich set of options to drive std::regex -- which include sub-patterns, back-tracking and what not, that may not be considered important today, but is a solid way to plan the future requirements. I am also uncomfortable in enhancing this little custom code in future, as it always comes up -- developers will easily break it, if not the original developer, who BTW, is very bright.

Therefore, once we outrun the usefulness of std::regex, I would be the first to suggest that we should write custom code -- now or later. I would apply that logic not just to std::regex, but also to other parts of the standard library, but need driven.

Thanks.

Reply:
For the specification - specify the constraints on your iterators (input? forward? random access?). Specify that `[start, last)` is a valid range (`first` and `last` are the strong convention). The value type must be char (or convertible to char?, or comparable with char?). What is the complexity (in operation counts)? What is the maximum recursion depth? Globbing typically excludes path separators, so note if that is not the case here. The old [SGI STL documentation](https://www.boost.org/sgi/stl/search.html) is a great example of solid specification practices.